powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/OperationServiceBehavior.java

```diff
@@ -716,6 +729,62 @@ private Map<String, String> mapMerge(Map<String, String> m1, Map<String, String>
         return m3;
     }
 
+    @SneakyThrows(GenericServiceException.class)
+    private void generateAndSetOtpToOperationDetail(final OperationEntity operation, final OperationDetailResponse operationDetailResponse) {
+        operationDetailResponse.setTotp(generateTotp(operation, powerAuthServiceConfiguration.getProximityCheckOtpLength()));
+    }
+
+    private static String generateTotp(final OperationEntity operation, final int otpLength) throws GenericServiceException {
+        final String seed = operation.getTotpSeed();
+        if (seed == null) {
+            return null;
+        }
+
+        try {
+            return new String(Totp.generateTotpSha256(Base64.getDecoder().decode(seed), LocalDateTime.now(), otpLength));
+        } catch (CryptoProviderException e) {
+            logger.error("Unable to generate OTP for operation ID: {}, user ID: {}", operation.getId(), operation.getUserId(), e);
+            throw new GenericServiceException(ServiceError.OPERATION_ERROR, e.getMessage(), e.getLocalizedMessage());
+        }
+    }
+
+    private static String generateTotpSeed(final OperationCreateRequest request, final OperationTemplateEntity template) throws GenericServiceException {
+        if (Boolean.FALSE.equals(request.getProximityCheckEnabled())) {
+            logger.debug("Proximity check is disabled in request from user ID: {}", request.getUserId());
+            return null;
+        } else if (Boolean.TRUE.equals(request.getProximityCheckEnabled()) || template.isProximityCheckEnabled()) {
+            logger.debug("Proximity check is enabled, generating TOTP seed for user ID: {}, templateName: {}", request.getUserId(), template.getTemplateName());
+            final KeyGenerator keyGenerator = new KeyGenerator();
+            try {
+                final byte[] seed = keyGenerator.generateRandomBytes(PROXIMITY_OTP_SEED_LENGTH);
+                return Base64.getEncoder().encodeToString(seed);
+            } catch (CryptoProviderException e) {
+                logger.error("Unable to generate proximity OTP seed for operation, user ID: {}", request.getUserId(), e);
+                throw new GenericServiceException(ServiceError.OPERATION_ERROR, e.getMessage(), e.getLocalizedMessage());
+            }
+        }
+        return null;
+    }
+
+    private static boolean proximityCheckPassed(final OperationEntity operation, final OperationApproveRequest request) {
+        if (operation.getTotpSeed() == null) {
+            return true;
+        }
+        final String otp = request.getAdditionalData().get("proximity_otp"); // TODO Lubos - constant, better name, is additional data correct?
+        if (otp == null) {
+            logger.info("Proximity check enabled for operation ID: {} but OTP not sent", operation.getId());
```

Comment:
Shouldn't this log level be a warning? It seems to be an unusual state.

Reply:
Sounds reasonable.